### ISSUE TITLE
More hysplit functionality

### ISSUE THREAD

**Issue by barronh** [barronh/pseudonetcdf#38]
Opening from sidebar in #31 

@bbakernoaa would like to see PNC support for cdump, pdump and tdump files from HYSPLIT.

Status:
1) I think I have a reader for tdump, but I'll have to dig it up.
2) pdump should be easy, but test file (`pdump_510`) doesn't seem to match description in #31 
3) cdump tbd

# PDUMP
pdump_510 is written in the Fortran "unformatted" format using big-endian, where each record is preceded by a 4-byte integer describing the number of data bytes to follow and concluded with a matching 4-byte integer. The first record has 32 bytes or 8 4-byte integers. As a show of good faith, I tinkered until I think I understood quite a bit.

```
python -c "import numpy as np; print(np.fromfile('pdump_510', dtype = '>i,>8i,>i,>i,>4i,>4f,>i,>i,>2i,>4f,>i,>i,>2i,>i,>i,>2i,>i,>i,>6i,>i,>i,>6i,>i', count = 1))"
[ (32, [1313295648, 95, 10, 16, 0, 0, 1, 3], 32, 32, [95, 10, 16, 0], [40.0, -90.0, 10.0, 0.0], 32, 24, [72, 20], [5.0, 5.0, 40.0, -90.0], 24, 8, [1, 100], 8, 8, [1, 1413829460], 8, 24, [95, 10, 16, 0, 0, 0], 24, 24, [95, 10, 16, 12, 0, 0], 24)]
```
My interpretation so far:
 - Record 1 (>i,>8i,>i) : Buf,Particle number,Year,Month,Day,Hour,Min,Sec?,?,Buf
 - Record 2 (>i,>4i,>4f,>i) : Buf,Year,Month,Day,Hour,Lat0,Lon0,?,?,Buf
 - Record 3 (>i,>2i,>4f,>i) : Buf,DX,DY,Lat0,Lon0,Buf
 - Record 4 (>i,>2i,>i): ?
 - Record 5 (>i,>2i,>i): ?
 - Record 6 (>i,>6i,>i) : Buf,Year,Month,Day,Hour,Min,Sec,Buf
 - Record 7 (>i,>6i,>i) : Buf,Year,Month,Day,Hour,Min,Sec,Buf

Basically, the definitions in issue #31 are either wrong or pdump_510 is not a good example and I could spend a lot of time trying to reverse engineer it.

@bbakernoaa - Can you clarify?

**Comment by bbakernoaa** [comment]
@barronh try this file
https://drive.google.com/open?id=1WLWdos7awpdGpWHafjUdSa87MT2xgCLY

**Comment by barronh** [comment]
@bbakernoaa I apparently need permissions.

**Comment by bbakernoaa** [comment]
Try again

**Comment by bbakernoaa** [comment]
@barronh I'm sorry but try these 
[hysplitfiles.tar.gz](https://github.com/barronh/pseudonetcdf/files/1682003/hysplitfiles.tar.gz)

**Comment by barronh** [comment]
@bbakernoaa - I have a working PARDUMP reader (arlpardump) with a working isMine. I am adding this in the develop branch. If you install from there, it should work.

I need testing and some input on a few things.

1. Some of the units should be updated (pls, help)

2. PARTICLE_MASS said `Particle pollutant mass (times the number of pollutants)`. Does that meant that I should expect to have an array of NPOLLUTANTS or that the mass is multiplied by the number of pollutants.

3. How does POLLUTANT in Rec2 sub rec 3 relate?

https://github.com/barronh/pseudonetcdf/blob/develop/src/PseudoNetCDF/noaafiles/_pdump.py

**Comment by bbakernoaa** [comment]
@barronh I'm going to get someone much more familiar with the inner workings of HYSPLIT to chime in here.  @amcz could you help out here?

**Comment by bbakernoaa** [comment]
@amcz ended up emailing me.  Here is what she had to say.

 1) Mass units are arbitrary 
    Height is in meters
    AGE - particle age in minutes (time since release).

 2) An array of N pollutants (each computational particle can represent multiple pollutant species. So the array would give the mass of each)

 3)  POLLUTANT in record 2 sub3 refers to  the pollutant index.  

There are two ways to model multiple pollutant species. The most common way (in my experience) is that each computational particle represents only
one particle species. The pardump file specifies which species is represented by the POLLUTANT id (shown in bold below). If this is done then the Number of Pollutants
in the header record is 1. 

The second way is to stick multiple pollutants onto the same computational particle. This is mainly useful if you need to do some kind of simple chemical transformation.
In this case Number of pollutants in the header file specifies how many species are on each computational particle.
And the pollutant mass will be an array of size number of pollutants which specifies the mass of each species on that computational particle.
POLLUTANT in Record is still the pollutant id but it will always be the same number.

### Record #1 (header record)
INT*4 Number of particles
INT*4 Number of pollutants
INT*4 Time of particle dump (YEAR, MONTH, DAY, HOUR, MINUTES)
 

### Record #2 - Loop to record: Number of particles
REAL*4 Particle pollutant mass (times the number of pollutants)
REAL*4 Particle LATITUDE, LONGITUDE, HEIGHT, SIGMA-U,SIGMA-V, SIGMA-X
INT*4 Particle AGE, DISTRIBUTION, **POLLUTANT**, METEO-GRID,SORT-INDEX

I hope this clears things up, let me know if you have any other questions.

**Comment by barronh** [comment]
@bbakernoaa  - I've updated the relative time to use the time stamp from the header. Please confirm. I also added tdump. All of these are somewhat experimental, so feedback is important.

**Comment by barronh** [comment]
This was actually an indirect problem that could have affected any of them... Not sure why it just affected one. It's been fixed and confirmed to work with xarray. I also made some other fixes to tdump.  Remember to view the results with a careful eye.
